### PR TITLE
Close response body when request done

### LINKED ISSUE #2730: Eventbus failed to send message when sending over 1000  msgs in 30 seconds
**What happened**:
I did a stress test on message router(eventbus -> rest) and found that when the edge sends more than 1000 messages in 30 seconds, the cloud cannot receive messages anymore.
```bash
E0325 17:27:04.249397   25154 eventbus.go:106] message is send to target fail. msgID: c43d1dd7-430e-4f62-b7ac-7fffae81e0b8, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
I0325 17:27:04.255001   25154 message.go:57] get a message, header:{ID:7b226c3e-59c9-4e4a-914d-30ea0d4af6e3 ParentID: Timestamp:1616664423507 ResourceVersion: Sync:false} router:{Source:bus Group:user Operation:upload Resource:node/centos78-edge-0/default/test-1}
I0325 17:27:04.255044   25154 message.go:57] get a message, header:{ID:56470d9a-d2b8-4818-89e4-3766a9f58387 ParentID: Timestamp:1616664423508 ResourceVersion: Sync:false} router:{Source:bus Group:user Operation:upload Resource:node/centos78-edge-0/default/test-1}
I0325 17:27:04.255068   25154 message.go:57] get a message, header:{ID:924ff1c8-5601-45e8-a776-085be2246f46 ParentID: Timestamp:1616664423509 ResourceVersion: Sync:false} router:{Source:bus Group:user Operation:upload Resource:node/centos78-edge-0/default/test-1}
I0325 17:27:04.255082   25154 message.go:57] get a message, header:{ID:4fbd3576-e8fe-40d0-bb18-dabef775722f ParentID: Timestamp:1616664423510 ResourceVersion: Sync:false} router:{Source:bus Group:user Operation:upload Resource:node/centos78-edge-0/default/test-1}
I0325 17:27:04.255130   25154 http.go:40] tlsConfig InsecureSkipVerify true
E0325 17:27:04.255209   25154 eventbus.go:106] message is send to target fail. msgID: 4fbd3576-e8fe-40d0-bb18-dabef775722f, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
I0325 17:27:04.255264   25154 http.go:40] tlsConfig InsecureSkipVerify true
E0325 17:27:04.255344   25154 eventbus.go:106] message is send to target fail. msgID: 7b226c3e-59c9-4e4a-914d-30ea0d4af6e3, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
I0325 17:27:04.255378   25154 http.go:40] tlsConfig InsecureSkipVerify true
E0325 17:27:04.255451   25154 eventbus.go:106] message is send to target fail. msgID: 56470d9a-d2b8-4818-89e4-3766a9f58387, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
I0325 17:27:04.255490   25154 http.go:40] tlsConfig InsecureSkipVerify true
E0325 17:27:04.255534   25154 eventbus.go:106] message is send to target fail. msgID: 924ff1c8-5601-45e8-a776-085be2246f46, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
```

**What you expected to happen**:
No limit for cloud

**How to reproduce it (as minimally and precisely as possible)**:
sending over 1000  msgs to the edge eventbus in 30 seconds

**Anything else we need to know?**:
I know that there is a ulimit to enlarge open files. But it always creates a new client for the connection. So why not disable the KeepAlive?
https://github.com/kubeedge/kubeedge/blob/f30b2021cded5191aa116f8ff51dbc9dc1f677d3/cloud/pkg/router/utils/http/http.go#L30-L35
https://github.com/kubeedge/kubeedge/blob/f30b2021cded5191aa116f8ff51dbc9dc1f677d3/cloud/pkg/router/provider/rest/rest.go#L173-L179

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.3
- KubeEdge version(e.g. `cloudcore --version` and `edgecore --version`): v1.6.0
- ulimit `open files` : 1024

### PR DESCRIPTION
**Pull request by JackZxj**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Increase the concurrency of sending messages from eventbus to rest:
```bash
# In edge node:
# It means publish "hello" to topic for 1050 times.
[root@centos78-edge-0 ~]# ./kubeedge-router-manager-bench -t er -b tcp://127.0.0.1:1883 -e "default/test-1" -m "hello" -n 1050
2021/03/26 18:36:13 eventbus.go:30: Sending start: 2021-03-26 10:36:13.857187449 +0000 UTC
2021/03/26 18:36:15 eventbus.go:44: Sending end: 2021-03-26 10:36:15.281538685 +0000 UTC
2021/03/26 18:36:15 eventbus.go:47: Test done.
Sum: 1050, Faild: 0, Time-Spending: 1.424350994s

# In cloud node:
# It means creates a listener and expects to receive message for 1050 times
[root@10 kubeedge-router-manager-bench]# ./kubeedge-router-manager-bench -t er -r 127.0.0.1:8080/test-1 -n 1050
2021/03/26 18:36:16 rest.go:65: Test done.
Sum: 1050, Faild: 0, Time-Spending: 1.422007239s
```
You can get this bench here: https://github.com/JackZxj/kubeedge-router-manager-bench

**Which issue(s) this PR fixes**:

Fixes #2730

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**: NONE

### CODE REVIEW

**Comment by fisherxu** [comment]
The dco check still failed @JackZxj Please follow the error detail to fix.

**Comment by JackZxj** [comment]
> The dco check still failed @JackZxj Please follow the error detail to fix.

I have fixed it. It is my first time to contribute this project, so I'm not familiar with these processes... @fisherxu

**Comment by fisherxu** [comment]
cc @liufen90

**Review by liufen90 — COMMENTED** [review]
the response is not a http response in some case.so can not modify it here.

**Comment by JackZxj** [comment]
https://github.com/kubeedge/kubeedge/blob/11f848536e48387218e86ee45814fde770287453/cloud/pkg/router/listener/message.go#L55-L63
Emm, but I found that only the router was referring to this function. I did a test on it and found that it could be `nil` or `http response`, what else could it be? 

In my test, this unclosed response did affect the number of messages received. How about this?
```golang
		if resp != nil {
			if v, ok := resp.(*http.Response); ok {
				if err = v.Body.Close(); err != nil {
					klog.Errorf("close response occur error, msgID: %s, reason: %s", message.GetID(), err.Error())
				}
			}
		}
```

File: cloud/pkg/router/listener/message.go [hunk 1]
```diff
@@ -76,10 +77,15 @@ func (mh *MessageHandler) HandleMessage(message *model.Message) error {
...
-		_, err := handler(message)
+		resp, err := handler(message)
 		if err != nil {
 			klog.Errorf("handle message occur error, msgID: %s, reason: %s", message.GetID(), err.Error())
 		}
+		if resp != nil {
+			if err = resp.(*http.Response).Body.Close(); err != nil {
+				klog.Errorf("close response occur error, msgID: %s, reason: %s", message.GetID(), err.Error())
+			}
+		}
```
Comment (WintonChan) [inline cloud/pkg/router/listener/message.go:88]: plz add it to func httpHandler in cloud\pkg\router\listener\http.go

Comment (fisherxu) [reply cloud/pkg/router/listener/message.go:88]: Has the comment been fixed?

**Comment by JackZxj** [comment]
Without offense, I'm just a new learner. If there's something wrong with me, please let me know.

@WintonChan Please have a check carefully. I think `cloud\pkg\router\listener\http.go` is for `rest->eventbus`

I did a test in your way, the max messages to send from eventbus to rest was 1012. It's as same as before:
```bash
# in cloud
# test tool:
[root@10 kubeedge-router-manager-bench]# ./kubeedge-router-manager-bench -t er -r 127.0.0.1:8080/test-1 -n 1050
2021/04/01 17:08:29 rest.go:61: It timed out 300 seconds.
Sum: 1050, Faild: 38, Time-Spending: 1.563620864s
# test log
E0401 17:10:25.858098   11469 eventbus.go:106] message is send to target fail. msgID: 03784ec1-2848-487e-80d9-db2920d23f3b, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
E0401 17:10:25.858196   11469 eventbus.go:106] message is send to target fail. msgID: c88ae80f-6df5-421c-9d98-9d10d8d738a4, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
...
```
but if modify by my way, it can do at least 1500:
```bash
[root@centos78-edge-0 ~]# ./kubeedge-router-manager-bench -t er -b tcp://127.0.0.1:1883 -e "default/test-1" -m "hello" -n 1500
2021/04/01 17:12:32 eventbus.go:30: Sending start: 2021-04-01 09:12:32.953671564 +0000 UTC
2021/04/01 17:12:34 eventbus.go:44: Sending end: 2021-04-01 09:12:34.739960057 +0000 UTC
2021/04/01 17:12:34 eventbus.go:47: Test done.
Sum: 1500, Faild: 0, Time-Spending: 1.786288384s
```

And @liufen90 , In my test, this response could be nil or http response, what else could it be?

Honestly, after my modification, there still will be an error of TOO MANY OPEN FILES if send too much message in one time. But I think what I do will increasing the concurrency.

If you two think it's better to modify `ulimit` directly, you can close pr.

**Comment by WintonChan** [comment]
> Without offense, I'm just a new learner. If there's something wrong with me, please let me know.
> 
> @WintonChan Please have a check carefully. I think `cloud\pkg\router\listener\http.go` is for `rest->eventbus`
> 
> I did a test in your way, the max messages to send from eventbus to rest was 1012. It's as same as before:
> 
> ```shell
> # in cloud
> # test tool:
> [root@10 kubeedge-router-manager-bench]# ./kubeedge-router-manager-bench -t er -r 127.0.0.1:8080/test-1 -n 1050
> 2021/04/01 17:08:29 rest.go:61: It timed out 300 seconds.
> Sum: 1050, Faild: 38, Time-Spending: 1.563620864s
> # test log
> E0401 17:10:25.858098   11469 eventbus.go:106] message is send to target fail. msgID: 03784ec1-2848-487e-80d9-db2920d23f3b, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
> E0401 17:10:25.858196   11469 eventbus.go:106] message is send to target fail. msgID: c88ae80f-6df5-421c-9d98-9d10d8d738a4, target: rest, err:Post "http://127.0.0.1:8080/test-1": dial tcp 127.0.0.1:8080: socket: too many open files
> ...
> ```
> 
> but if modify by my way, it can do at least 1500:
> 
> ```shell
> # in edge
> [root@centos78-edge-0 ~]# ./kubeedge-router-manager-bench -t er -b tcp://127.0.0.1:1883 -e "default/test-1" -m "hello" -n 1500
> 2021/04/01 17:12:32 eventbus.go:30: Sending start: 2021-04-01 09:12:32.953671564 +0000 UTC
> 2021/04/01 17:12:34 eventbus.go:44: Sending end: 2021-04-01 09:12:34.739960057 +0000 UTC
> 2021/04/01 17:12:34 eventbus.go:47: Test done.
> Sum: 1500, Faild: 0, Time-Spending: 1.786288384s
> 
> # in cloud
> [root@10 kubeedge-router-manager-bench]# ./kubeedge-router-manager-bench -t er -r 127.0.0.1:8080/test-1 -n 1500 -timeout 60
> 2021/04/01 17:12:35 rest.go:68: Test done.
> Sum: 1500, Faild: 0, Time-Spending: 1.790345691s
> ```
> 
> And @liufen90 , In my test, this response could be nil or http response, what else could it be?
> 
> Honestly, after my modification, there still will be an error of TOO MANY OPEN FILES if send too much message in one time. But I think what I do will increasing the concurrency.
> 
> If you two think it's better to modify `ulimit` directly, you can close pr.

I'm sorry to reply you so late. I have check it my code carefully again.  Please ignore my previous comment. In this case , rest to eventbus ,the response of handle will be ignored and  should be closed when its type is http response. Thanks for your test.  

And @liufen90 , In my test, this response could be nil or http response, what else could it be?
Currently, it is only nil or http response

File: cloud/pkg/router/listener/message.go (hunk 1 above)
Comment (WintonChan) [reply cloud/pkg/router/listener/message.go:88]: plz ignore this comment.

**Comment by JackZxj** [comment]
So... How about this PR? @WintonChan

**Review by WintonChan — APPROVED** [review]

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve
Merged now, please fix any issue in follow-up PRs.

**Pull request merged**